## Notebook 1 – Exploratory Data Analysis and Feature Selection

### Purpose

This notebook performs the initial preparation and exploratory analysis of the 2016 Toronto Neighbourhood Profile dataset.

The objectives of this notebook are to:

- import and inspect the raw dataset;
- evaluate data quality and completeness;
- identify candidate socioeconomic indicators;
- define the dependent (target) and independent (predictor) variables;
- perform feature selection and justification;
- generate a machine-learning-ready dataset for subsequent analyses.

This notebook addresses the first stage of the predictive analytics workflow and provides the foundation for the relationship analysis, dimensionality reduction, clustering, and predictive modelling notebooks.

In [217]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from ydata_profiling import ProfileReport

pd.set_option("display.max_columns", None)

print("Libraries imported successfully.")

Libraries imported successfully.


## Section 2 – Project Input & Output 
### Input

**Dataset**

- Toronto Neighbourhood Profile 2016 (City of Toronto Open Data)
- File: `neighbourhood-profiles-2016-140-model.csv`

The dataset contains demographic, socioeconomic, education, housing, labour market, and population characteristics for all 140 Toronto neighbourhoods.

### Output

This notebook will produce:

- A cleaned and structured machine-learning-ready dataset
- Selected socioeconomic indicators for predictive modelling
- Data quality assessment
- Feature selection justification
- Exploratory data analysis (EDA)
- YData Profiling report
- Exported dataset: `neighbourhood_profiles_2016_ml_ready.csv`

### Data Loading

In [218]:
file_path = "../../data/raw/neighbourhood-profiles-2016-140-model.csv"

df_raw = pd.read_csv(file_path)

print("Dataset loaded successfully.")
print(f"Rows: {df_raw.shape[0]}")
print(f"Columns: {df_raw.shape[1]}")

Dataset loaded successfully.
Rows: 2383
Columns: 146


### Standardize Column Names

The original Toronto Neighbourhood Profile dataset contains column names that differ slightly across dataset versions (*Characteristic* versus *Attribute*).

To maintain consistency throughout this project and ensure that the 2011 and 2016 notebooks follow the same structure, selected column names are standardized immediately after loading the dataset.

Standardizing column names improves code readability, simplifies feature extraction, and allows reusable code across multiple datasets.

In [219]:
df_raw.rename(
    columns={
        "Characteristic": "Attribute"},
    inplace=True)

print("Column names standardized successfully.")

Column names standardized successfully.


## Section 3 – Initial Data Inspection

Before performing data cleaning and feature selection, the raw dataset is examined to understand its overall structure, dimensions, data types, and content.

The objectives of this section are to:

- examine the size and structure of the dataset;
- inspect the first few records;
- identify variable names and data types;
- detect potential inconsistencies or formatting issues;
- establish a baseline understanding of the data before preprocessing.

This initial inspection provides the foundation for the subsequent data quality assessment and feature selection process.

In [220]:
print("=" * 60)
print("Dataset Shape")
print("=" * 60)
print(df_raw.shape)

print("\n")

print("=" * 60)
print("First Five Rows")
print("=" * 60)
display(df_raw.head())

print("\n")

print("=" * 60)
print("Column Names")
print("=" * 60)
display(pd.DataFrame(df_raw.columns, columns=["Column Name"]))

Dataset Shape
(2383, 146)


First Five Rows


,_id,Category,Topic,Data Source,Attribute,City of Toronto,Agincourt North,Agincourt South-Malvern West,Alderwood,Annex,Banbury-Don Mills,Bathurst Manor,Bay Street Corridor,Bayview Village,Bayview Woods-Steeles,Bedford Park-Nortown,Beechborough-Greenbrook,Bendale,Birchcliffe-Cliffside,Black Creek,Blake-Jones,Briar Hill-Belgravia,Bridle Path-Sunnybrook-York Mills,Broadview North,Brookhaven-Amesbury,Cabbagetown-South St. James Town,Caledonia-Fairbank,Casa Loma,Centennial Scarborough,Church-Yonge Corridor,Clairlea-Birchmount,Clanton Park,Cliffcrest,Corso Italia-Davenport,Danforth,Danforth East York,Don Valley Village,Dorset Park,Dovercourt-Wallace Emerson-Junction,Downsview-Roding-CFB,Dufferin Grove,East End-Danforth,Edenbridge-Humber Valley,Eglinton East,Elms-Old Rexdale,Englemount-Lawrence,Eringate-Centennial-West Deane,Etobicoke West Mall,Flemingdon Park,Forest Hill North,Forest Hill South,Glenfield-Jane Heights,Greenwood-Coxwell,Guildwood,Henry Farm,High Park North,High Park-Swansea,Highland Creek,Hillcrest Village,Humber Heights-Westmount,Humber Summit,Humbermede,Humewood-Cedarvale,Ionview,Islington-City Centre West,Junction Area,Keelesdale-Eglinton West,Kennedy Park,Kensington-Chinatown,Kingsview Village-The Westway,Kingsway South,Lambton Baby Point,L'Amoreaux,Lansing-Westgate,Lawrence Park North,Lawrence Park South,Leaside-Bennington,Little Portugal,Long Branch,Malvern,Maple Leaf,Markland Wood,Milliken,Mimico (includes Humber Bay Shores),Morningside,Moss Park,Mount Dennis,Mount Olive-Silverstone-Jamestown,Mount Pleasant East,Mount Pleasant West,New Toronto,Newtonbrook East,Newtonbrook West,Niagara,North Riverdale,North St. James Town,Oakridge,Oakwood Village,O'Connor-Parkview,Old East York,Palmerston-Little Italy,Parkwoods-Donalda,Pelmo Park-Humberlea,Playter Estates-Danforth,Pleasant View,Princess-Rosethorn,Regent Park,Rexdale-Kipling,Rockcliffe-Smythe,Roncesvalles,Rosedale-Moore Park,Rouge,Runnymede-Bloor West Village,Rustic,Scarborough Village,South Parkdale,South Riverdale,St.Andrew-Windfields,Steeles,Stonegate-Queensway,Tam O'Shanter-Sullivan,Taylor-Massey,The Beaches,Thistletown-Beaumond Heights,Thorncliffe Park,Trinity-Bellwoods,University,Victoria Village,Waterfront Communities-The Island,West Hill,West Humber-Clairville,Westminster-Branson,Weston,Weston-Pelham Park,Wexford/Maryvale,Willowdale East,Willowdale West,Willowridge-Martingrove-Richview,Woburn,Woodbine Corridor,Woodbine-Lumsden,Wychwood,Yonge-Eglinton,Yonge-St.Clair,York University Heights,Yorkdale-Glen Park
0,1,Neighbourhood Information,Neighbourhood Information,City of Toronto,Neighbourhood Number,NaN,129,128,20,95,42,34,76,52,49,39,112,127,122,24,69,108,41,57,30,71,109,96,133,75,120,33,123,92,66,59,47,126,93,26,83,62,9,138,5,32,11,13,44,102,101,25,65,140,53,88,87,134,48,8,21,22,106,125,14,90,110,124,78,6,15,114,117,38,105,103,56,84,19,132,29,12,130,17,135,73,115,2,99,104,18,50,36,82,68,74,121,107,54,58,80,45,23,67,46,10,72,4,111,86,98,131,89,28,139,85,70,40,116,16,118,61,63,3,55,81,79,43,77,136,1,35,113,91,119,51,37,7,137,64,60,94,100,97,27,31
1,2,Neighbourhood Information,Neighbourhood Information,City of Toronto,TSNS2020 Designation,NaN,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,NIA,No Designation,No Designation,NIA,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,Emerging Neighbourhood,No Designation,NIA,No Designation,No Designation,No Designation,NIA,NIA,Emerging Neighbourhood,No Designation,No Designation,NIA,No Designation,No Designation,NIA,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,No Designation,Emerging Neighbourhood,NIA,NIA,No Designation,NIA,No Designation,No Designation,NIA,NIA,No Desig



Column Names


,Column Name
0,_id
1,Category
2,Topic
3,Data Source
4,Attribute
...,...
141,Wychwood
142,Yonge-Eglinton
143,Yonge-St.Clair
144,York University Heights


### Section 4 – Data Quality Assessment

Before selecting variables for predictive modelling, the quality of the dataset must be assessed to ensure that the data are complete, consistent, and suitable for analysis.

The objectives of this section are to:

- identify missing values;
- detect duplicate observations;
- examine data types;
- identify invalid or inconsistent values;
- evaluate the overall completeness of the dataset.

This assessment provides an initial understanding of the dataset's quality before more detailed profiling and feature selection are performed.

In [221]:
print("=" * 60)
print("Dataset Information")
print("=" * 60)

df_raw.info()

Dataset Information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2383 entries, 0 to 2382
Columns: 146 entries, _id to Yorkdale-Glen Park
dtypes: int64(1), object(145)
memory usage: 2.7+ MB


### Missing Value Assessment

Missing values can reduce data quality and negatively affect exploratory analysis and predictive modelling.

This assessment identifies variables containing missing values and quantifies the proportion of missing observations within the dataset.

Understanding missingness at this stage helps determine whether variables should be retained, transformed, or excluded during feature selection.

In [222]:
missing_values = (
    df_raw.isnull()
          .sum()
          .to_frame(name="Missing Values"))

missing_values["Missing (%)"] = (
    missing_values["Missing Values"] / len(df_raw) * 100
).round(2)

missing_values = missing_values.sort_values(
    by="Missing Values",
    ascending=False)

display(missing_values.head(20))

,Missing Values,Missing (%)
Lansing-Westgate,56,2.35
Lawrence Park North,56,2.35
North Riverdale,56,2.35
North St. James Town,56,2.35
Oakridge,56,2.35
Oakwood Village,56,2.35
O'Connor-Parkview,56,2.35
Old East York,56,2.35
Palmerston-Little Italy,56,2.35
Parkwoods-Donalda,56,2.35


In [223]:
rows_with_missing = df_raw[df_raw.isnull().any(axis=1)]

print(f"Number of indicator rows containing missing values: {len(rows_with_missing)}")

display(
    rows_with_missing[
        ["Category", "Topic", "Attribute"]    ])

Number of indicator rows containing missing values: 62


,Category,Topic,Attribute
0,Neighbourhood Information,Neighbourhood Information,Neighbourhood Number
1,Neighbourhood Information,Neighbourhood Information,TSNS2020 Designation
946,Income,Income of individuals in 2015,Median total income in 2015 among recipien...
948,Income,Income of individuals in 2015,Median after-tax income in 2015 among reci...
951,Income,Income of individuals in 2015,Median market income in 2015 among recipie...
...,...,...,...
1960,Journey to work,Commuting destination,Total - Commuting destination for the employed...
1961,Journey to work,Commuting destination,Commute within census subdivision (CSD) of r...
1962,Journey to work,Commuting destination,Commute to a different census subdivision (C...
1963,Journey to work,Commuting destination,Commute to a different census subdivision (C...


### Interpretation

A total of 62 indicator rows contain one or more missing values.

Most missing values occur in indicators that are not measured for every neighbourhood or represent aggregate statistics rather than neighbourhood-level observations.

The presence of missing values will be considered during feature selection, and only variables with sufficient neighbourhood-level coverage will be retained for predictive modelling.

### Section 5 – Candidate Feature Assessment

#### Objective

Before feature selection, candidate socioeconomic indicators are systematically reviewed to determine their suitability for predictive modelling.

Each candidate variable is assessed based on three criteria:

- Availability at the neighbourhood level
- Consistency with the 2011 Toronto Neighbourhood Profile
- Relevance to neighbourhood socioeconomic vulnerability

Only variables that satisfy these criteria will be selected for the final machine learning dataset.

A helper function is used to search the dataset and verify each candidate variable before feature extraction.

In [224]:
def find_indicator(keyword):
    """
    Search the Attribute column for a keyword.
    """
    results = df_raw[
        df_raw["Attribute"].str.contains(keyword, case=False, na=False)  ]

    return results[["Category", "Topic", "Attribute"]]

### 5.1 Population

In [225]:
find_indicator("Population, 2016")

,Category,Topic,Attribute
2,Population,Population and dwellings,"Population, 2016"


### 5.2 Labour Force Indicators

In [226]:
find_indicator("Participation rate")

,Category,Topic,Attribute
1888,Labour,Labour force status,Participation rate
1896,Labour,Labour force status,Participation rate (Males)
1904,Labour,Labour force status,Participation rate (Females)


In [227]:
find_indicator("Employment rate")

,Category,Topic,Attribute
1889,Labour,Labour force status,Employment rate
1890,Labour,Labour force status,Unemployment rate
1897,Labour,Labour force status,Employment rate (Males)
1898,Labour,Labour force status,Unemployment rate (Males)
1905,Labour,Labour force status,Employment rate (Females)
1906,Labour,Labour force status,Unemployment rate (Females)


In [228]:
find_indicator("Unemployment rate")

,Category,Topic,Attribute
1890,Labour,Labour force status,Unemployment rate
1898,Labour,Labour force status,Unemployment rate (Males)
1906,Labour,Labour force status,Unemployment rate (Females)


### 5.3 Educational Attainment

In [229]:
find_indicator("Bachelor")

,Category,Topic,Attribute
1708,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...
1709,Education,"Highest certificate, diploma or degree","University certificate, diploma or degree ..."
1710,Education,"Highest certificate, diploma or degree",Bachelor's degree
1711,Education,"Highest certificate, diploma or degree",University certificate or diploma above ...
1723,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...
1724,Education,"Highest certificate, diploma or degree","University certificate, diploma or degree ..."
1725,Education,"Highest certificate, diploma or degree",Bachelor's degree
1726,Education,"Highest certificate, diploma or degree",University certificate or diploma above ...


In [230]:
find_indicator("No certificate")

,Category,Topic,Attribute
1703,Education,"Highest certificate, diploma or degree","No certificate, diploma or degree"
1715,Education,"Highest certificate, diploma or degree","No certificate, diploma or degree"


### 5.4 Immigration

In [231]:
find_indicator("Immigrants")

,Category,Topic,Attribute
1151,Immigration and citizenship,Immigrant status and period of immigration,Non-immigrants
1152,Immigration and citizenship,Immigrant status and period of immigration,Immigrants
1296,Immigration and citizenship,Admission category and applicant type,Economic immigrants
1299,Immigration and citizenship,Admission category and applicant type,Immigrants sponsored by family
1301,Immigration and citizenship,Admission category and applicant type,Other immigrants


### 5.5 Visible Minority

In [232]:
find_indicator("Visible")

,Category,Topic,Attribute
1335,Visible minority,Visible minority population,Total - Visible minority for the population in...
1336,Visible minority,Visible minority population,Total visible minority population
1346,Visible minority,Visible minority population,Visible minority; n.i.e.
1347,Visible minority,Visible minority population,Multiple visible minorities
1348,Visible minority,Visible minority population,Not a visible minority


### 5.6 Housing Indicators

In [233]:
find_indicator("Rent")

,Category,Topic,Attribute
94,"Families, households and marital status",Family characteristics,Total lone-parent families by sex of parent
95,"Families, households and marital status",Family characteristics,Female parent
96,"Families, households and marital status",Family characteristics,Male parent
102,"Families, households and marital status",Family characteristics,Lone-parent census families in private households
1081,Income,Income of economic families in 2015,Total - Income statistics in 2015 for lone-par...
1088,Income,Income of economic families in 2015,Median total income of lone-parent economic ...
1089,Income,Income of economic families in 2015,Median after-tax income of lone-parent econo...
1090,Income,Income of economic families in 2015,Average family size of lone-parent economic ...
1101,Income,Income of economic families in 2015,Total - Income statistics in 2015 for lone-par...
1102,Income,Income of economic families in 2015,Average total income of lone-parent economic...


In [234]:
find_indicator("Core housing")

,Category,Topic,Attribute
1681,Housing,Core housing need,Rate of core housing need
1682,Housing,Core housing need,Number of households in core housing need
1684,Housing,Core housing need,"Core housing need, Unsuitable only"
1685,Housing,Core housing need,"Core housing need, Inadequate only"
1686,Housing,Core housing need,"Core housing need, Unaffordable only"
1687,Housing,Core housing need,"Core housing need, two or more dimensions"


In [235]:
find_indicator("Unaffordable")

,Category,Topic,Attribute
1686,Housing,Core housing need,"Core housing need, Unaffordable only"
1688,Housing,Core housing need,Rate of unaffordable housing
1689,Housing,Core housing need,Unaffordable housing not in core need
1690,Housing,Core housing need,Unaffordable housing in core need
1691,Housing,Core housing need,Unaffordable housing not applicable for core need
1692,Housing,Core housing need,"Rate of unaffordable housing, owner-households"
1693,Housing,Core housing need,"Rate of unaffordable housing, renter-households"


### 5.7 Family Structure

In [236]:
find_indicator("lone-parent")

,Category,Topic,Attribute
94,"Families, households and marital status",Family characteristics,Total lone-parent families by sex of parent
102,"Families, households and marital status",Family characteristics,Lone-parent census families in private households
1081,Income,Income of economic families in 2015,Total - Income statistics in 2015 for lone-par...
1088,Income,Income of economic families in 2015,Median total income of lone-parent economic ...
1089,Income,Income of economic families in 2015,Median after-tax income of lone-parent econo...
1090,Income,Income of economic families in 2015,Average family size of lone-parent economic ...
1101,Income,Income of economic families in 2015,Total - Income statistics in 2015 for lone-par...
1102,Income,Income of economic families in 2015,Average total income of lone-parent economic...
1103,Income,Income of economic families in 2015,Average after-tax income of lone-parent econ...


### 5.8 Target Variable

In [237]:
find_indicator("low")

,Category,Topic,Attribute
1120,Income,Low income in 2015,Total - Low-income status in 2015 for the popu...
1125,Income,Low income in 2015,"In low income based on the Low-income measure,..."
1130,Income,Low income in 2015,Prevalence of low income based on the Low-inco...
1133,Income,Low income in 2015,In low income based on the Low-income cut-offs...
1139,Income,Low income in 2015,Prevalence of low income based on the Low-inco...
1708,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...
1723,Education,"Highest certificate, diploma or degree",University certificate or diploma below ba...
2304,Income,Income sources,Guaranteed Income Supplement (GIS) and s...
2305,Income,Income sources,Guaranteed Income Supplement (GIS) and s...
2307,Income,Income sources,Guaranteed Income Supplement (GIS) and s...


In [238]:
find_indicator("after-tax")

,Category,Topic,Attribute
947,Income,Income of individuals in 2015,Number of after-tax income recipients aged 1...
948,Income,Income of individuals in 2015,Median after-tax income in 2015 among reci...
960,Income,Income of individuals in 2015,Number of after-tax income recipients aged 1...
961,Income,Income of individuals in 2015,Average after-tax income in 2015 among rec...
984,Income,Income of individuals in 2015,Total - After-tax income groups in 2015 for th...
985,Income,Income of individuals in 2015,Without after-tax income
986,Income,Income of individuals in 2015,With after-tax income
987,Income,Income of individuals in 2015,Percentage with after-tax income
1019,Income,Income of households in 2015,Median after-tax income of households in 201...
1022,Income,Income of households in 2015,Median after-tax income of one-person ho...


### Interpratation

The candidate feature assessment confirmed that multiple neighbourhood-level socioeconomic indicators are available within the 2016 Toronto Neighbourhood Profile.

Each candidate variable was reviewed to verify its availability, completeness, and relevance for predictive modelling. Variables representing labour market conditions, educational attainment, immigration, visible minority population, housing conditions, family structure, and neighbourhood-level low-income prevalence were successfully identified.

The assessment also confirmed that some indicators available in earlier Census Profiles, such as median individual income and median after-tax income, are not consistently reported at the neighbourhood level in the 2016 dataset. Consequently, only variables with complete neighbourhood-level coverage will be retained for the final modelling dataset.

The verified indicators will now be organized into the final feature set for extraction and transformation.

### Section 6 – Final Feature Selection

#### Objective

Following the candidate feature assessment, the final set of socioeconomic indicators is selected for predictive modelling.

Variables were retained based on four criteria:

- availability at the neighbourhood level;
- completeness and quality of the data;
- consistency with the 2011 Toronto Neighbourhood Profile where possible;
- relevance to neighbourhood socioeconomic vulnerability.

The selected variables consist of one target variable and multiple predictor variables that will be extracted and transformed into the final machine learning dataset.

In [239]:
feature_mapping = {

    "Population": 2,

    "Participation_Rate": 1888,
    "Employment_Rate": 1889,
    "Unemployment_Rate": 1890,

    "Bachelors_Degree": 1710,
    "No_Diploma": 1704,

    "Immigrant_Population": 1152,
    "Visible_Minority": 1336,

    "Renters": 1629,
    "Core_Housing_Need": 1682,      
    "Unaffordable_Housing": 1688,

    "Lone_Parent_Families": 102,

    "Low_Income_AfterTax_Pct": 1130}

In [240]:
feature_table = (
    pd.DataFrame(
        {
            "Research Variable": feature_mapping.keys(),
            "Source Row": feature_mapping.values()
        }   ))

display(feature_table)

,Research Variable,Source Row
0,Population,2
1,Participation_Rate,1888
2,Employment_Rate,1889
3,Unemployment_Rate,1890
4,Bachelors_Degree,1710
5,No_Diploma,1704
6,Immigrant_Population,1152
7,Visible_Minority,1336
8,Renters,1629
9,Core_Housing_Need,1682


#### Interpretation

The final feature set consists of one target variable and fifteen predictor variables representing key dimensions of neighbourhood socioeconomic conditions.

The selected predictors capture labour market participation, income, educational attainment, immigration, visible minority status, housing conditions, family structure, and population characteristics.

These variables provide a comprehensive representation of neighbourhood vulnerability while maintaining compatibility with the 2011 Toronto Neighbourhood Profile for the predictive modelling stage.

The following section extracts the selected indicators and constructs the modelling dataset.

### Section 7 – Feature Extraction
#### Objective

Using the finalized feature mapping, the corresponding indicators are extracted from the 2016 Toronto Neighbourhood Profile.

Only the selected socioeconomic variables required for predictive modelling are retained. The resulting dataset will serve as the foundation for the machine learning-ready neighbourhood dataset.

### 7.1 Feature Extraction

In [241]:
selected_df = df_raw.loc[
    list(feature_mapping.values())
].copy()

print("=" * 60)
print("FEATURE EXTRACTION SUMMARY")
print("=" * 60)

print(f"\nSelected variables: {selected_df.shape[0]}")
print(f"Columns: {selected_df.shape[1]}")

display(
    selected_df[
        ["_id", "Category", "Topic", "Attribute"]   ])

FEATURE EXTRACTION SUMMARY

Selected variables: 13
Columns: 146


,_id,Category,Topic,Attribute
2,3,Population,Population and dwellings,"Population, 2016"
1888,1889,Labour,Labour force status,Participation rate
1889,1890,Labour,Labour force status,Employment rate
1890,1891,Labour,Labour force status,Unemployment rate
1710,1711,Education,"Highest certificate, diploma or degree",Bachelor's degree
1704,1705,Education,"Highest certificate, diploma or degree",Secondary (high) school diploma or equivalen...
1152,1153,Immigration and citizenship,Immigrant status and period of immigration,Immigrants
1336,1337,Visible minority,Visible minority population,Total visible minority population
1629,1630,Housing,Household characteristics,Renter
1682,1683,Housing,Core housing need,Number of households in core housing need


### 7.2 Remove Metadata

The selected indicators still contain descriptive metadata used for documentation purposes.

For machine learning, only the neighbourhood-level values are required. Therefore, the metadata columns are removed while preserving the selected socioeconomic indicators.

In [242]:
model_data = selected_df.iloc[:, 4:].copy()

print("=" * 60)
print("MODEL DATA")
print("=" * 60)

print(model_data.shape)

display(model_data.head())

MODEL DATA
(13, 142)


,Attribute,City of Toronto,Agincourt North,Agincourt South-Malvern West,Alderwood,Annex,Banbury-Don Mills,Bathurst Manor,Bay Street Corridor,Bayview Village,Bayview Woods-Steeles,Bedford Park-Nortown,Beechborough-Greenbrook,Bendale,Birchcliffe-Cliffside,Black Creek,Blake-Jones,Briar Hill-Belgravia,Bridle Path-Sunnybrook-York Mills,Broadview North,Brookhaven-Amesbury,Cabbagetown-South St. James Town,Caledonia-Fairbank,Casa Loma,Centennial Scarborough,Church-Yonge Corridor,Clairlea-Birchmount,Clanton Park,Cliffcrest,Corso Italia-Davenport,Danforth,Danforth East York,Don Valley Village,Dorset Park,Dovercourt-Wallace Emerson-Junction,Downsview-Roding-CFB,Dufferin Grove,East End-Danforth,Edenbridge-Humber Valley,Eglinton East,Elms-Old Rexdale,Englemount-Lawrence,Eringate-Centennial-West Deane,Etobicoke West Mall,Flemingdon Park,Forest Hill North,Forest Hill South,Glenfield-Jane Heights,Greenwood-Coxwell,Guildwood,Henry Farm,High Park North,High Park-Swansea,Highland Creek,Hillcrest Village,Humber Heights-Westmount,Humber Summit,Humbermede,Humewood-Cedarvale,Ionview,Islington-City Centre West,Junction Area,Keelesdale-Eglinton West,Kennedy Park,Kensington-Chinatown,Kingsview Village-The Westway,Kingsway South,Lambton Baby Point,L'Amoreaux,Lansing-Westgate,Lawrence Park North,Lawrence Park South,Leaside-Bennington,Little Portugal,Long Branch,Malvern,Maple Leaf,Markland Wood,Milliken,Mimico (includes Humber Bay Shores),Morningside,Moss Park,Mount Dennis,Mount Olive-Silverstone-Jamestown,Mount Pleasant East,Mount Pleasant West,New Toronto,Newtonbrook East,Newtonbrook West,Niagara,North Riverdale,North St. James Town,Oakridge,Oakwood Village,O'Connor-Parkview,Old East York,Palmerston-Little Italy,Parkwoods-Donalda,Pelmo Park-Humberlea,Playter Estates-Danforth,Pleasant View,Princess-Rosethorn,Regent Park,Rexdale-Kipling,Rockcliffe-Smythe,Roncesvalles,Rosedale-Moore Park,Rouge,Runnymede-Bloor West Village,Rustic,Scarborough Village,South Parkdale,South Riverdale,St.Andrew-Windfields,Steeles,Stonegate-Queensway,Tam O'Shanter-Sullivan,Taylor-Massey,The Beaches,Thistletown-Beaumond Heights,Thorncliffe Park,Trinity-Bellwoods,University,Victoria Village,Waterfront Communities-The Island,West Hill,West Humber-Clairville,Westminster-Branson,Weston,Weston-Pelham Park,Wexford/Maryvale,Willowdale East,Willowdale West,Willowridge-Martingrove-Richview,Woburn,Woodbine Corridor,Woodbine-Lumsden,Wychwood,Yonge-Eglinton,Yonge-St.Clair,York University Heights,Yorkdale-Glen Park
2,"Population, 2016","2,731,571","29,113","23,757","12,054","30,526","27,695","15,873","25,797","21,396","13,154","23,236","6,577","29,960","22,291","21,737","7,727","14,257","9,266","11,499","17,757","11,669","9,955","10,968","13,362","31,340","26,984","16,472","15,935","14,133","9,666","17,180","27,051","25,003","36,625","35,052","11,785","21,381","15,535","22,776","9,456","22,372","18,588","11,848","21,933","12,806","10,732","30,491","14,417","9,917","15,723","22,162","23,925","12,494","16,934","10,948","12,416","15,545","14,365","13,641","43,965","14,366","11,058","17,123","17,945","22,000","9,271","7,985","43,993","16,164","14,607","15,179","16,828","15,559","10,084","43,794","10,111","10,554","26,572","33,964","17,455","20,506","13,593","32,954","16,775","29,658","11,463","16,097","23,831","31,180","11,916","18,615","13,845","21,210","18,675","9,233","13,826","34,805","10,722","7,804","15,818","11,051","10,803","10,529","22,246","14,974","20,923","46,496","10,070","9,941","16,724","21,849","27,876","17,812","24,623","25,051","27,446","15,683","21,567","10,360","21,108","16,556","7,607","17,510","65,913","27,392","33,312","26,274","17,992","11,098","27,917","50,434","16,936","22,156","53,485","12,541","7,865","14,349","11,817","12,528","27,593","14,804"
1888,Participation rate,64.7,55.4,59,66.5,70.6,59.9,65,62.7,63.4,55.7,63.8,59.5,58,65.4,54.8,65,67.6,58.3,66.5,63.5,71.3,65.9,67.2,64.2,75.1,65.3,67.9,61.9,68.1,68.5,67,61,63.3,70.3,61.7,72.4,70.1,64.4,59.2,60.7,63,63.6,65.7,56.

In [243]:
model_data = model_data.drop(columns=["Attribute"])

print(model_data.shape)

display(model_data.head())

(13, 141)


,City of Toronto,Agincourt North,Agincourt South-Malvern West,Alderwood,Annex,Banbury-Don Mills,Bathurst Manor,Bay Street Corridor,Bayview Village,Bayview Woods-Steeles,Bedford Park-Nortown,Beechborough-Greenbrook,Bendale,Birchcliffe-Cliffside,Black Creek,Blake-Jones,Briar Hill-Belgravia,Bridle Path-Sunnybrook-York Mills,Broadview North,Brookhaven-Amesbury,Cabbagetown-South St. James Town,Caledonia-Fairbank,Casa Loma,Centennial Scarborough,Church-Yonge Corridor,Clairlea-Birchmount,Clanton Park,Cliffcrest,Corso Italia-Davenport,Danforth,Danforth East York,Don Valley Village,Dorset Park,Dovercourt-Wallace Emerson-Junction,Downsview-Roding-CFB,Dufferin Grove,East End-Danforth,Edenbridge-Humber Valley,Eglinton East,Elms-Old Rexdale,Englemount-Lawrence,Eringate-Centennial-West Deane,Etobicoke West Mall,Flemingdon Park,Forest Hill North,Forest Hill South,Glenfield-Jane Heights,Greenwood-Coxwell,Guildwood,Henry Farm,High Park North,High Park-Swansea,Highland Creek,Hillcrest Village,Humber Heights-Westmount,Humber Summit,Humbermede,Humewood-Cedarvale,Ionview,Islington-City Centre West,Junction Area,Keelesdale-Eglinton West,Kennedy Park,Kensington-Chinatown,Kingsview Village-The Westway,Kingsway South,Lambton Baby Point,L'Amoreaux,Lansing-Westgate,Lawrence Park North,Lawrence Park South,Leaside-Bennington,Little Portugal,Long Branch,Malvern,Maple Leaf,Markland Wood,Milliken,Mimico (includes Humber Bay Shores),Morningside,Moss Park,Mount Dennis,Mount Olive-Silverstone-Jamestown,Mount Pleasant East,Mount Pleasant West,New Toronto,Newtonbrook East,Newtonbrook West,Niagara,North Riverdale,North St. James Town,Oakridge,Oakwood Village,O'Connor-Parkview,Old East York,Palmerston-Little Italy,Parkwoods-Donalda,Pelmo Park-Humberlea,Playter Estates-Danforth,Pleasant View,Princess-Rosethorn,Regent Park,Rexdale-Kipling,Rockcliffe-Smythe,Roncesvalles,Rosedale-Moore Park,Rouge,Runnymede-Bloor West Village,Rustic,Scarborough Village,South Parkdale,South Riverdale,St.Andrew-Windfields,Steeles,Stonegate-Queensway,Tam O'Shanter-Sullivan,Taylor-Massey,The Beaches,Thistletown-Beaumond Heights,Thorncliffe Park,Trinity-Bellwoods,University,Victoria Village,Waterfront Communities-The Island,West Hill,West Humber-Clairville,Westminster-Branson,Weston,Weston-Pelham Park,Wexford/Maryvale,Willowdale East,Willowdale West,Willowridge-Martingrove-Richview,Woburn,Woodbine Corridor,Woodbine-Lumsden,Wychwood,Yonge-Eglinton,Yonge-St.Clair,York University Heights,Yorkdale-Glen Park
2,"2,731,571","29,113","23,757","12,054","30,526","27,695","15,873","25,797","21,396","13,154","23,236","6,577","29,960","22,291","21,737","7,727","14,257","9,266","11,499","17,757","11,669","9,955","10,968","13,362","31,340","26,984","16,472","15,935","14,133","9,666","17,180","27,051","25,003","36,625","35,052","11,785","21,381","15,535","22,776","9,456","22,372","18,588","11,848","21,933","12,806","10,732","30,491","14,417","9,917","15,723","22,162","23,925","12,494","16,934","10,948","12,416","15,545","14,365","13,641","43,965","14,366","11,058","17,123","17,945","22,000","9,271","7,985","43,993","16,164","14,607","15,179","16,828","15,559","10,084","43,794","10,111","10,554","26,572","33,964","17,455","20,506","13,593","32,954","16,775","29,658","11,463","16,097","23,831","31,180","11,916","18,615","13,845","21,210","18,675","9,233","13,826","34,805","10,722","7,804","15,818","11,051","10,803","10,529","22,246","14,974","20,923","46,496","10,070","9,941","16,724","21,849","27,876","17,812","24,623","25,051","27,446","15,683","21,567","10,360","21,108","16,556","7,607","17,510","65,913","27,392","33,312","26,274","17,992","11,098","27,917","50,434","16,936","22,156","53,485","12,541","7,865","14,349","11,817","12,528","27,593","14,804"
1888,64.7,55.4,59,66.5,70.6,59.9,65,62.7,63.4,55.7,63.8,59.5,58,65.4,54.8,65,67.6,58.3,66.5,63.5,71.3,65.9,67.2,64.2,75.1,65.3,67.9,61.9,68.1,68.5,67,61,63.3,70.3,61.7,72.4,70.1,64.4,59.2,60.7,63,63.6,65.7,56.5,70.9,65.8,55,69.1,59.5,65.2,72.2,73.3,60.9,53.

### 7.3 Rename Research Variables

The extracted indicators are relabelled using standardized research variable names.

Replacing the original dataset row identifiers with descriptive variable names improves readability and ensures consistency throughout the subsequent data preparation and predictive modelling stages.

In [244]:
model_data.index = feature_mapping.keys()

print("=" * 60)
print("RESEARCH VARIABLE NAMES")
print("=" * 60)

display(model_data.head())

RESEARCH VARIABLE NAMES


,City of Toronto,Agincourt North,Agincourt South-Malvern West,Alderwood,Annex,Banbury-Don Mills,Bathurst Manor,Bay Street Corridor,Bayview Village,Bayview Woods-Steeles,Bedford Park-Nortown,Beechborough-Greenbrook,Bendale,Birchcliffe-Cliffside,Black Creek,Blake-Jones,Briar Hill-Belgravia,Bridle Path-Sunnybrook-York Mills,Broadview North,Brookhaven-Amesbury,Cabbagetown-South St. James Town,Caledonia-Fairbank,Casa Loma,Centennial Scarborough,Church-Yonge Corridor,Clairlea-Birchmount,Clanton Park,Cliffcrest,Corso Italia-Davenport,Danforth,Danforth East York,Don Valley Village,Dorset Park,Dovercourt-Wallace Emerson-Junction,Downsview-Roding-CFB,Dufferin Grove,East End-Danforth,Edenbridge-Humber Valley,Eglinton East,Elms-Old Rexdale,Englemount-Lawrence,Eringate-Centennial-West Deane,Etobicoke West Mall,Flemingdon Park,Forest Hill North,Forest Hill South,Glenfield-Jane Heights,Greenwood-Coxwell,Guildwood,Henry Farm,High Park North,High Park-Swansea,Highland Creek,Hillcrest Village,Humber Heights-Westmount,Humber Summit,Humbermede,Humewood-Cedarvale,Ionview,Islington-City Centre West,Junction Area,Keelesdale-Eglinton West,Kennedy Park,Kensington-Chinatown,Kingsview Village-The Westway,Kingsway South,Lambton Baby Point,L'Amoreaux,Lansing-Westgate,Lawrence Park North,Lawrence Park South,Leaside-Bennington,Little Portugal,Long Branch,Malvern,Maple Leaf,Markland Wood,Milliken,Mimico (includes Humber Bay Shores),Morningside,Moss Park,Mount Dennis,Mount Olive-Silverstone-Jamestown,Mount Pleasant East,Mount Pleasant West,New Toronto,Newtonbrook East,Newtonbrook West,Niagara,North Riverdale,North St. James Town,Oakridge,Oakwood Village,O'Connor-Parkview,Old East York,Palmerston-Little Italy,Parkwoods-Donalda,Pelmo Park-Humberlea,Playter Estates-Danforth,Pleasant View,Princess-Rosethorn,Regent Park,Rexdale-Kipling,Rockcliffe-Smythe,Roncesvalles,Rosedale-Moore Park,Rouge,Runnymede-Bloor West Village,Rustic,Scarborough Village,South Parkdale,South Riverdale,St.Andrew-Windfields,Steeles,Stonegate-Queensway,Tam O'Shanter-Sullivan,Taylor-Massey,The Beaches,Thistletown-Beaumond Heights,Thorncliffe Park,Trinity-Bellwoods,University,Victoria Village,Waterfront Communities-The Island,West Hill,West Humber-Clairville,Westminster-Branson,Weston,Weston-Pelham Park,Wexford/Maryvale,Willowdale East,Willowdale West,Willowridge-Martingrove-Richview,Woburn,Woodbine Corridor,Woodbine-Lumsden,Wychwood,Yonge-Eglinton,Yonge-St.Clair,York University Heights,Yorkdale-Glen Park
Population,"2,731,571","29,113","23,757","12,054","30,526","27,695","15,873","25,797","21,396","13,154","23,236","6,577","29,960","22,291","21,737","7,727","14,257","9,266","11,499","17,757","11,669","9,955","10,968","13,362","31,340","26,984","16,472","15,935","14,133","9,666","17,180","27,051","25,003","36,625","35,052","11,785","21,381","15,535","22,776","9,456","22,372","18,588","11,848","21,933","12,806","10,732","30,491","14,417","9,917","15,723","22,162","23,925","12,494","16,934","10,948","12,416","15,545","14,365","13,641","43,965","14,366","11,058","17,123","17,945","22,000","9,271","7,985","43,993","16,164","14,607","15,179","16,828","15,559","10,084","43,794","10,111","10,554","26,572","33,964","17,455","20,506","13,593","32,954","16,775","29,658","11,463","16,097","23,831","31,180","11,916","18,615","13,845","21,210","18,675","9,233","13,826","34,805","10,722","7,804","15,818","11,051","10,803","10,529","22,246","14,974","20,923","46,496","10,070","9,941","16,724","21,849","27,876","17,812","24,623","25,051","27,446","15,683","21,567","10,360","21,108","16,556","7,607","17,510","65,913","27,392","33,312","26,274","17,992","11,098","27,917","50,434","16,936","22,156","53,485","12,541","7,865","14,349","11,817","12,528","27,593","14,804"
Participation_Rate,64.7,55.4,59,66.5,70.6,59.9,65,62.7,63.4,55.7,63.8,59.5,58,65.4,54.8,65,67.6,58.3,66.5,63.5,71.3,65.9,67.2,64.2,75.1,65.3,67.9,61.9,68.1,68.5,67,61,63.3,70.3,61.7,72.4,70.1,64.4,59.2,60.7,63,63.6,65.7,56.5,70.9,65.8,55,69.1,59.5,

### 7.4 Transpose the Dataset

The selected indicators are currently stored as rows and neighbourhoods as columns.

To prepare the data for predictive modelling, the dataset is transposed so that each neighbourhood becomes one observation and each socioeconomic indicator becomes a feature.

This produces the standard machine learning format where:

- rows represent neighbourhoods;
- columns represent predictor variables and the target variable.

In [245]:
ml_ready = model_data.T.copy()

print("=" * 60)
print("TRANSPOSED DATASET")
print("=" * 60)

print(ml_ready.shape)

display(ml_ready.head())

TRANSPOSED DATASET
(141, 13)


,Population,Participation_Rate,Employment_Rate,Unemployment_Rate,Bachelors_Degree,No_Diploma,Immigrant_Population,Visible_Minority,Renters,Core_Housing_Need,Unaffordable_Housing,Lone_Parent_Families,Low_Income_AfterTax_Pct
City of Toronto,"2,731,571",64.7,59.3,8.2,534610,561090,"1,266,005","1,385,855","525,835","240,780",36.6,"152,595",20.2
Agincourt North,"29,113",55.4,50,9.8,4380,7460,"19,995","26,365","1,720","2,220",34.7,1650,21.3
Agincourt South-Malvern West,"23,757",59,53.2,9.8,4210,6090,"15,095","20,155","2,250","1,795",38.2,1360,22.6
Alderwood,"12,054",66.5,62.4,6.1,1660,2960,"3,970","2,490",950,540,21.7,670,8.2
Annex,"30,526",70.6,65.8,6.7,9135,4270,"8,270","7,270","9,870","2,420",40.4,895,18.3


### 7.5 Create the Neighbourhood Identifier

Following the transposition, neighbourhood names are stored as the DataFrame index.

The index is converted into a dedicated column named **Neighbourhood**, allowing it to function as the unique identifier for each observation during subsequent analysis and modelling.

In [246]:
ml_ready.reset_index(inplace=True)

ml_ready.rename(
    columns={"index": "Neighbourhood"},
    inplace=True)

print("=" * 60)
print("NEIGHBOURHOOD IDENTIFIER CREATED")
print("=" * 60)

display(ml_ready.head())

NEIGHBOURHOOD IDENTIFIER CREATED


,Neighbourhood,Population,Participation_Rate,Employment_Rate,Unemployment_Rate,Bachelors_Degree,No_Diploma,Immigrant_Population,Visible_Minority,Renters,Core_Housing_Need,Unaffordable_Housing,Lone_Parent_Families,Low_Income_AfterTax_Pct
0,City of Toronto,"2,731,571",64.7,59.3,8.2,534610,561090,"1,266,005","1,385,855","525,835","240,780",36.6,"152,595",20.2
1,Agincourt North,"29,113",55.4,50,9.8,4380,7460,"19,995","26,365","1,720","2,220",34.7,1650,21.3
2,Agincourt South-Malvern West,"23,757",59,53.2,9.8,4210,6090,"15,095","20,155","2,250","1,795",38.2,1360,22.6
3,Alderwood,"12,054",66.5,62.4,6.1,1660,2960,"3,970","2,490",950,540,21.7,670,8.2
4,Annex,"30,526",70.6,65.8,6.7,9135,4270,"8,270","7,270","9,870","2,420",40.4,895,18.3


### 7.6 Convert Variables to Numeric

Neighbourhood identifiers remain as text, while all socioeconomic indicators are converted to numeric values.

This ensures compatibility with statistical analyses, machine learning algorithms, and data visualization techniques.

In [247]:
for column in ml_ready.columns[1:]:
    ml_ready[column] = (
        ml_ready[column]
        .astype(str)
        .str.replace(",", "", regex=False)    )

    ml_ready[column] = pd.to_numeric(
        ml_ready[column],
        errors="coerce"    )

print("=" * 60)
print("DATA TYPES")
print("=" * 60)

display(ml_ready.dtypes)

DATA TYPES


Neighbourhood               object
Population                   int64
Participation_Rate         float64
Employment_Rate            float64
Unemployment_Rate          float64
Bachelors_Degree             int64
No_Diploma                   int64
Immigrant_Population         int64
Visible_Minority             int64
Renters                      int64
Core_Housing_Need            int64
Unaffordable_Housing       float64
Lone_Parent_Families         int64
Low_Income_AfterTax_Pct    float64
dtype: object

### 7.7 Data Quality Assessment

Prior to exporting the modelling dataset, several quality assurance checks are performed.

The dataset is assessed for:

- missing values;
- duplicate neighbourhood records;
- overall dataset structure.

These validation steps ensure that the machine learning dataset is complete and suitable for predictive modelling.

In [248]:
print("=" * 60)
print("FINAL DATASET SHAPE")
print("=" * 60)

print(ml_ready.shape)

FINAL DATASET SHAPE
(141, 14)


## 8. Export Machine Learning Dataset

## Purpose

Following feature engineering, restructuring, and data validation, the finalized machine learning-ready dataset is exported for use in the subsequent stages of the project.

The exported dataset serves as the standardized analytical input for all remaining notebooks, ensuring that every statistical analysis, dimensionality reduction technique, clustering procedure, and predictive modelling experiment is performed using a consistent and reproducible dataset.

The exported file contains:

- **141 observations** (City of Toronto and 140 neighbourhoods);
- **14 predictor variables** representing demographic, socioeconomic, housing, labour market, education, and population characteristics;
- **1 target variable** 
representing neighbourhood-level low-income prevalence (After-Tax Low Income Measure)

1 identifier: Neighbourhood

14 predictor variables (X):
Population
Participation_Rate
Employment_Rate
Unemployment_Rate
Median_Income
Median_AfterTax_Income
Bachelors_Degree
No_Diploma
Immigrant_Population
Visible_Minority
Renters
Core_Housing_Need
Unaffordable_Housing
Lone_Parent_Families

1 target variable (Y):
Low_Income_AfterTax_Pct

Exporting the dataset at this stage establishes a reproducible analytical workflow and provides a single authoritative data source for the remainder of the capstone project.

In [249]:
output_path = "../../data/processed/neighbourhood_profiles_2016_ml_ready.csv"

ml_ready.to_csv(
    output_path,
    index=False)

print("=" * 60)
print("EXPORT COMPLETE")
print("=" * 60)

print(f"Dataset saved to:\n{output_path}")

print("\nDataset Shape:")
print(ml_ready.shape)

display(ml_ready.head())

EXPORT COMPLETE
Dataset saved to:
../../data/processed/neighbourhood_profiles_2016_ml_ready.csv

Dataset Shape:
(141, 14)


,Neighbourhood,Population,Participation_Rate,Employment_Rate,Unemployment_Rate,Bachelors_Degree,No_Diploma,Immigrant_Population,Visible_Minority,Renters,Core_Housing_Need,Unaffordable_Housing,Lone_Parent_Families,Low_Income_AfterTax_Pct
0,City of Toronto,2731571,64.7,59.3,8.2,534610,561090,1266005,1385855,525835,240780,36.6,152595,20.2
1,Agincourt North,29113,55.4,50.0,9.8,4380,7460,19995,26365,1720,2220,34.7,1650,21.3
2,Agincourt South-Malvern West,23757,59.0,53.2,9.8,4210,6090,15095,20155,2250,1795,38.2,1360,22.6
3,Alderwood,12054,66.5,62.4,6.1,1660,2960,3970,2490,950,540,21.7,670,8.2
4,Annex,30526,70.6,65.8,6.7,9135,4270,8270,7270,9870,2420,40.4,895,18.3


## 9. Automated Data Profiling

### Purpose

Following the export of the finalized machine learning-ready dataset, an automated exploratory data analysis (EDA) is performed using **YData Profiling** to independently validate the exported analytical dataset.

Unlike the manual exploratory analysis conducted earlier in this notebook, automated profiling provides a comprehensive assessment of the exported dataset by generating descriptive statistics, variable distributions, correlation analysis, missing value diagnostics, duplicate detection, and data quality warnings.

This step serves as an additional quality assurance procedure to verify that the exported dataset is complete, internally consistent, and suitable for subsequent statistical analyses and predictive modelling.

The generated HTML report is stored in the project's **reports** directory and forms supplementary documentation for this study.

In [250]:
from ydata_profiling import ProfileReport

In [251]:
validation_df = pd.read_csv(output_path)

profile = ProfileReport(
    validation_df,
    title="Toronto Neighbourhood Profile (2016) - Machine Learning Dataset",
    explorative=True)

print("=" * 60)
print("YDATA PROFILING")
print("=" * 60)
print("Profiling report generated successfully.")

YDATA PROFILING
Profiling report generated successfully.


In [252]:
report_path = "../../reports/Toronto_ML_Profile_Report_2016.html"

profile.to_file(report_path)

print("=" * 60)
print("REPORT SAVED")
print("=" * 60)
print(f"Location:\n{report_path}")


Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

100%|██████████| 14/14 [00:00<00:00, 259.27it/s]


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

REPORT SAVED
Location:
../../reports/Toronto_ML_Profile_Report_2016.html


### Interpretation

The automated YData Profiling report successfully validated the exported machine learning-ready dataset.

The report confirmed the dataset structure, variable distributions, descriptive statistics, missing values, duplicate observations, and correlation patterns.

The generated HTML report provides supplementary documentation supporting the manual exploratory data analysis and confirms that the exported dataset is suitable for subsequent statistical analyses and predictive modelling.

## 10. Conclusion

### Summary

This notebook completed the data preparation and exploratory data analysis (EDA) phase of the project by transforming the raw 2016 Toronto Neighbourhood Profile dataset into a structured machine learning-ready dataset suitable for statistical analysis and predictive modelling.

The original dataset required substantial preprocessing due to its non-standard format, where socioeconomic indicators were stored as rows and neighbourhoods as columns. A manual exploratory data analysis was therefore conducted to identify, extract, and standardize the socioeconomic indicators relevant to the research objectives.

Following feature selection, the dataset was restructured into a conventional analytical format consisting of neighbourhoods as observations and socioeconomic indicators as variables. Data quality was assessed through manual validation procedures and an automated YData Profiling report, confirming that the final dataset contained the expected variable structure, appropriate data types, and no missing values within the selected indicators.

The final machine learning-ready dataset consists of:

- **141 observations** (City of Toronto and 140 neighbourhoods);
- **12 socioeconomic predictor variables**;
- **1 target variable** representing neighbourhood-level low-income prevalence based on the After-Tax Low Income Measure (LIM-AT).

The finalized dataset has been exported for use in the subsequent stages of this study.

---

## Next Steps

The next notebook investigates the statistical relationships between the selected socioeconomic indicators and neighbourhood-level low-income prevalence.

Research Question 1 will be addressed by quantifying the strength and direction of the relationships between predictor variables and the target variable using descriptive statistics, Pearson correlation analysis, Spearman rank correlation analysis, feature ranking, and multicollinearity assessment (Variance Inflation Factor).

The results of these analyses will identify the socioeconomic indicators most strongly associated with neighbourhood vulnerability and provide the statistical foundation for the predictive modelling developed in the subsequent notebooks.